azure-openai[major]: Integration of Azure OpenAI SDK with Langchain (#3919)
langchain-ai/langchainjs: libs/langchain-azure-openai/README.md

This package contains the Azure SDK for OpenAI LangChain.js integrations.

It provides Azure OpenAI support through the [Azure SDK for OpenAI](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/openai/openai) library. 

## Installation

```bash npm2yarn
npm install @langchain/azure-openai
```

This package, along with the main LangChain package, depends on [`@langchain/core`](https://npmjs.com/package/@langchain/core/).
If you are using this package with other LangChain packages, you should make sure that all of the packages depend on the same instance of @langchain/core.
You can do so by adding appropriate fields to your project's `package.json` like this:

```json
{
  "name": "your-project",
  "version": "0.0.0",
  "dependencies": {
    "@langchain/openai": "^0.0.9",
    "langchain": "0.0.207"
  },
  "resolutions": {
    "@langchain/core": "0.1.5"
  },
  "overrides": {
    "@langchain/core": "0.1.5"
  },
  "pnpm": {
    "overrides": {
      "@langchain/core": "0.1.5"
    }
  }
}
```

The field you need depends on the package manager you're using, but we recommend adding a field for the common `yarn`, `npm`, and `pnpm` to maximize compatibility.

## Chat Models

> This package contains the `AzureChatOpenAI` class, which is the recommended way to interface with the OpenAI series of models.

Reviewer: I think we should highlight that this is for Azure:
```suggestion
This package contains the `AzureChatOpenAI` class, which is the recommended way to interface with deployed models on Azure OpenAI.
```

Author: Done

To use, install the requirements, and configure your environment.

```bash
export AZURE_OPENAI_API_ENDPOINT=<your_endpoint>
export AZURE_OPENAI_API_KEY=<your_key>
export AZURE_OPENAI_API_DEPLOYMENT_NAME=<your_deployment_name>
```

Then initialize the model and make the calls:

```typescript
import { AzureChatOpenAI } from "@langchain/azure-openai";

const model = new AzureChatOpenAI({
  modelName: "gpt-4-1106-preview",
  // Note that the following are optional, and will default to the values below
  // if not provided.
  azureOpenAIEndpoint: process.env.AZURE_OPENAI_API_ENDPOINT,
  azureOpenAIApiKey: process.env.AZURE_OPENAI_API_KEY,

>   azureOpenAIApiDeploymentName: process.env.AZURE_OPENAI_API_DEPLOYMENT_NAME,

Reviewer: I wonder why do we need to provide a model name? Azure exposes a specific model through a specific deployment, should this be renamed to deployment name?

Author: The model name is not used by @azure/openai package. But, this is needed for internal functionalities of langchain. Also, this is consistent with the langchain/openai package

});
const response = await model.invoke(new HumanMessage("Hello world!"));
```

### Streaming

```typescript
import { AzureChatOpenAI } from "@langchain/azure-openai";

const model = new AzureChatOpenAI({
  modelName: "gpt-4-1106-preview",
  // Note that the following are optional, and will default to the values below
  // if not provided.
  azureOpenAIEndpoint: process.env.AZURE_OPENAI_API_ENDPOINT,
  azureOpenAIApiKey: process.env.AZURE_OPENAI_API_KEY,
  azureOpenAIApiDeploymentName: process.env.AZURE_OPENAI_API_DEPLOYMENT_NAME,
});
const response = await model.stream(new HumanMessage("Hello world!"));
```

## Embeddings

> This package also adds support for OpenAI's embeddings model.

Reviewer: ```suggestion
This package also supports embeddings with Azure OpenAI.
```

Author: Done
